projects/optic/src/utils/spec-loaders.ts

```diff
@@ -2,26 +2,146 @@ import fs from 'node:fs/promises';
 import {
-  ParseResult,
-  parseSpecVersion,
-  specFromInputToResults,
-} from '@useoptic/optic-ci/build/cli/commands/utils';
-import { validateOpenApiV3Document } from '@useoptic/openapi-io';
-import { filePathToGitPath } from '@useoptic/openapi-io';
+  validateOpenApiV3Document,
+  filePathToGitPath,
+  JsonSchemaSourcemap,
+  parseOpenAPIFromRepoWithSourcemap,
+  ParseOpenAPIResult,
+  parseOpenAPIWithSourcemap,
+} from '@useoptic/openapi-io';
+import { OpticCliConfig, VCS } from '../config';
+import { resolveGitRef } from './git-utils';
 
 const exec = promisify(callbackExec);
 
-export type { ParseResult };
+export type ParseResult = ParseOpenAPIResult & {
+  isEmptySpec: boolean;
+  context: {
+    vcs: 'git';
+    sha: string;
+  } | null;
+};
+
+enum SpecVersionFrom {
+  file,
+  git,
+  empty,
+}
+
+type SpecFromInput =
+  | {
+      from: SpecVersionFrom.file;
+      filePath: string;
+    }
+  | {
+      from: SpecVersionFrom.git;
+      branch: string;
+      name: string;
+    }
+  | {
+      from: SpecVersionFrom.empty;
+      value: OpenAPIV3.Document;
+    };
+
+function parseSpecVersion(
+  raw: string | undefined,
+  defaultSpec: OpenAPIV3.Document
+): SpecFromInput {
+  if (raw) {
+    if (raw.includes(':') && !(raw.startsWith('C:') || raw.startsWith('D:'))) {
+      const index = raw.indexOf(':');
+      const rev = raw.substring(0, index);
+      const name = raw.substring(index + 1);
+
+      return {
+        from: SpecVersionFrom.git,
+        name: name.startsWith('/') ? name.substring(1) : name,
+        branch: rev,
+      };
+    } else {
+      return {
+        from: SpecVersionFrom.file,
+        filePath: raw,
+      };
+    }
+  } else {
+    return {
+      from: SpecVersionFrom.empty,
+      value: defaultSpec,
+    };
+  }
+}
+
+async function specFromInputToResults(
+  input: SpecFromInput,
+  config: OpticCliConfig,
+  workingDir: string = process.cwd()
+): Promise<ParseResult> {
+  switch (input.from) {
+    case SpecVersionFrom.empty: {
+      const emptySpecName = 'empty.json';
+      const jsonLike = {
+        ...input.value,
+        ['x-optic-ci-empty-spec']: true,
+      } as OpenAPIV3.Document;
+      const sourcemap = new JsonSchemaSourcemap(emptySpecName);
+      await sourcemap.addFileIfMissingFromContents(
+        emptySpecName,
+        JSON.stringify(jsonLike, null, 2),
+        0
+      );
+
+      return {
+        jsonLike,
+        sourcemap,
+        isEmptySpec: true,
+        context: null,
+      };
+    }
+    case SpecVersionFrom.git: {
+      if (config.vcs?.type !== VCS.Git) {
+        throw new Error(`${workingDir} is not a git repo`);
+      }
+      return {
+        ...(await parseOpenAPIFromRepoWithSourcemap(
+          input.name,
+          config.root,
+          input.branch
+        )),
+        isEmptySpec: false,
+        context: {
+          vcs: 'git',
+          sha: await resolveGitRef(input.branch),
```

Comment:
We need to resolve the git ref since we can pass in branch names here. i.e. branch names will be converted into the shas they represent